### PR TITLE
Update cloud function spec with secret management and remove tenant config from versioning

### PR DESCRIPTION
**Pull request by carmenlau**
- Add secret management detail
    - Make secret immutable and define the detail behaviour
- Remove tenant config from versioning
    - Discussed with @Steven-Chan, both of us think since we only have single auth gear version, so it makes sense that app only have single version tenant config.

### CODE REVIEW

**Comment by carmenlau** [comment]
@kaojohnny @chpapa @Steven-Chan I update the spec to add some detail about secret management of CF. Please take a look when have time and let me know for any comments. I will start developing first. thanks!:)

File: features/268-next-cf/api.md [hunk 1]
```diff
@@ -614,3 +614,53 @@ Currently, micro-service allow developers to do routing in the app, while others
...
+
+To avoid human error, deployment will be failed if config file is using secrets
+that are not exist. Removing or renaming a secret will not affect existing deployment.
+If skygear user want to update a secret (e.g. key rotation), he will need to rename the
+existing secret to another name. Create a secret with the same name and re-deploy the functions.
+
+### Example: Key rotation
+
+```sh
+$ skycli secret rename MY_SECRET_KEY MY_SECRET_KEY_OLD
```
Comment (chpapa) [inline features/268-next-cf/api.md]: Will we provide a function to rotate key in future? A recommended way to handle it as a separate feature issue?

**Comment by carmenlau** [comment]
> What is our current secrets management model like? Is that each tenant can only access their own secrets?

Yes, secrets are under an app (tenant). Skygear user needs to add the secret to the app first by using`skycli secret create ...` and then they can use it in their cf config like this
```
cf:
  function1:
    type: http-handler
    path: /hello-world
    env: node
    src: js/hellow-world
    secrets:
      - DATABASE_URL
```

> We can keep it simple now but may need to jot down the attack vectors (e.g. user could possibly break something and access the k8s secrets store?) And create feature issues related with it.

In the current architecture, user will not access the k8s directly. Instead all the operations will be done through the skygear controller, so the attack vectors will may.
- If the k8s credentials leak and user can access the k8s cluster directly, all secrets may be exposed
- (From k8s) If a node is compromised, the only secrets that could potentially be exposed should be the secrets belonging to containers scheduled onto it
- (From k8s) If the master is compromised, all secrets in the cluster may be exposed

Since we are using managed k8s, the master, node and secret store security should be handled by provider. So different providers will be different. For google, *By default, GKE encrypts customer content stored at rest, including Secrets*. 

k8s secrets: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/secrets.md#constraints-and-assumptions
google gke encrypting secrets: https://cloud.google.com/kubernetes-engine/docs/how-to/encrypting-secrets

> Will we provide a function to rotate key in future? A recommended way to handle it as a separate feature issue?

Do you mean a single command for key rotation? Or you think this flow is not struct forward? Sure, we can create separate feature issue if needed.:)

**Comment by carmenlau** [comment]
Thanks for comments:) @chpapa 

> What is our current secrets management model like? Is that each tenant can only access their own secrets?

Yes, secrets are under an app (tenant). Skygear user needs to add the secret to the app first by using`skycli secret create ...` and then they can use it in their cf config like this
```
cf:
  function1:
    type: http-handler
    path: /hello-world
    env: node
    src: js/hellow-world
    secrets:
      - DATABASE_URL
```

> We can keep it simple now but may need to jot down the attack vectors (e.g. user could possibly break something and access the k8s secrets store?) And create feature issues related with it.

In the current architecture, user will not access the k8s directly. Instead all the operations will be done through the skygear controller, so the attack vectors will may.
- If the k8s credentials leak and user can access the k8s cluster directly, all secrets may be exposed
- (From k8s) If a node is compromised, the only secrets that could potentially be exposed should be the secrets belonging to containers scheduled onto it
- (From k8s) If the master is compromised, all secrets in the cluster may be exposed

Since we are using managed k8s, the master, node and secret store security should be handled by provider. So different providers will be different. For google, *By default, GKE encrypts customer content stored at rest, including Secrets*. 

k8s secrets: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/auth/secrets.md#constraints-and-assumptions
google gke encrypting secrets: https://cloud.google.com/kubernetes-engine/docs/how-to/encrypting-secrets

> Will we provide a function to rotate key in future? A recommended way to handle it as a separate feature issue?

Do you mean a single command for key rotation? Or you think this flow is not struct forward? Sure, we can create separate feature issue if needed.:)

**Comment by chpapa** [comment]
> In the current architecture, user will not access the k8s directly. Instead all the operations will be done through the skygear controller, so the attack vectors will may.

Yea, I just want to make sure we don't have stupid things like, etcd store all secrets, and it is open access, so anyone who can "guess/scan" the etcd node IP address could get all secrets.

 

> Do you mean a single command for key rotation? Or you think this flow is not struct forward? Sure, we can create separate feature issue if needed.:)

Yea I think it is not very straight forward, but could be a future issue.

**Comment by carmenlau** [comment]
@chpapa I created separate feature issue for key rotation. fyi https://github.com/SkygearIO/features/issues/301

Let me know if there is any further comments? I am going to merge it if it looks good:)

Any comments? @Steven-Chan @kaojohnny 

Thanks!! 🙏🏻

File: features/268-next-cf/api.md [hunk 2]
```diff
@@ -614,3 +614,53 @@ Currently, micro-service allow developers to do routing in the app, while others
...
+
+With immutable deployment, secrets are also immutable. For managing secrets,
+following operations are allowed.
+
+- skycli secret create
+- skycli secret remove
+- skycli secret rename
+
+To avoid human error, deployment will be failed if config file is using secrets
+that are not exist. Removing or renaming a secret will not affect existing deployment.
```
Comment (kaojohnny) [inline features/268-next-cf/api.md]: > Removing or renaming a secret will not affect existing deployment.

Want to confirming about rename secret.

1. After rename secret, all existing deployment keeps no change, a developer may need to update each `secrets` section of a `cf` in `skygear.yaml` and run deploy again to use the secret by the updated name.
2. When a developer run `skycli secret list`, he cannot see secrets with the previous name.

**Comment by carmenlau** [comment]
I am going to merge this feature issue and the secret feature is implemented. Let me know or open new feature issue if there is any problem.

**Pull request merged**